### ISSUE TITLE
perf: vmap over molecules in objective function

### ISSUE THREAD

**Issue by ericchansen** [ericchansen/q2mm#180]
## Summary

Explore `vmap` over molecules at the objective function level — evaluating all molecules in a single vectorised call rather than a Python loop. This is distinct from batching Hessians at the engine level (#176 Phase 1); it means making the entire per-molecule evaluation (energy → Hessian → eigenvalues → frequencies → residuals) a pure JAX function that can be vmapped.

## Motivation

The current objective function evaluates molecules sequentially in Python:

```python
for ref in self.reference.values:
    if mol_idx not in calc_cache:
        calc_cache[mol_idx] = self._evaluate_molecule(mol_idx, ff)
```

This is the worst-case pattern for GPU utilisation: many tiny kernel launches with Python overhead between each. `vmap` at the objective level would let XLA see all molecules at once, enabling maximum kernel fusion.

## Challenges

- Molecules have different atom counts → need padding or bucketing
- The evaluation includes eigenvalue decomposition (`jnp.linalg.eigh`) which must work under `vmap`
- Reference data (QM frequencies, weights) must be batched correspondingly
- The objective must be expressible as a pure JAX function (no Python control flow)

## Relationship to other approaches

This overlaps significantly with #176 (end-to-end JIT loss). In practice, `vmap` over molecules might be implemented as part of the end-to-end JIT loss function, or it could be done incrementally — first `vmap` the Hessian computation (#176 Phase 1), then extend `vmap` upward to cover eigenvalues and frequency extraction, and finally encompass the full residual computation.

## Related

- #176 (end-to-end differentiable loss — superset of this work)
- #178 (float32 viability — `vmap` benefits more from float32 due to larger batch sizes)
- #179 (JAX-ReaxFF — uses `vmap` over all geometries in one call)
